lib/Listener/UserDeletedListener.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright 2020 Christoph Wurst <christoph@winzerhof-wurst.at>
+ *
+ * @author 2020 Christoph Wurst <christoph@winzerhof-wurst.at>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace OCA\Mail\Listener;
+
+use OCA\Mail\Exception\ClientException;
+use OCA\Mail\Service\AccountService;
+use OCP\EventDispatcher\Event;
+use OCP\EventDispatcher\IEventListener;
+use OCP\ILogger;
+use OCP\User\Events\BeforeUserDeletedEvent;
+
+class UserDeletedListener implements IEventListener {
+
+	/** @var AccountService */
+	private $accountService;
+
+	/** @var ILogger */
+	private $logger;
+
+	public function __construct(AccountService $accountService,
+								ILogger $logger) {
+		$this->accountService = $accountService;
+		$this->logger = $logger;
+	}
+
+	public function handle(Event $event): void {
+		if (!($event instanceof BeforeUserDeletedEvent)) {
+			// Unrelated
+			return;
+		}
+
+		$user = $event->getUser();
+		foreach ($this->accountService->findByUserId($user->getUID()) as $account) {
```

Comment:
Would it make sense to move this to a background job?
I'm just thinking about the case where a user gets deleted but has a few huge mailboxes. Could still take some time for this all to get done then.

Reply:
Yeah, once we actually do the full cleanup right away. My current plan is to do it lazily in a background job.

Reply:
In that case a design consideration could be to do this for non-existing users rather than marking deleted users? This way, it will retroactively also clean up those that were deleted before the fix and other potential future database inconsistencies.